README.md
Spotiguess
==========

This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).

# Features:

- Connect to Spotify API to get your collaborative play-lists
- Select a play-list
- Start a gaming session with a join key (e.g. NWAQTX)
    + this device will be the game master
    + Can start game, kick players and act as an admin
    + (Defines number of rounds played)
- Up to 4 players can join the session using the key
    + Mobile devices, tablets or smart-phones
    + Desktop devices also possible
- Random song is selected from play-list
    + Information presented: album cover, artist, title and sample music
- The player can select any collaborator of the play-list as answer.
- Points are awarded for correct answers.
- A new random song is selected from play-list
- If number of rounds is reached a player point summary is displayed

# Technical details

- Node js back-end with express
- Socket.io for game-master <-> player communication
- Bootstrap for responsive design

# Extension features

- Meta questions:
    + Who added the most songs
    + Who has the longest streak (in songs)
    + Who has the longest streak (in time)
    + Who has added the most unavailable songs (from the perspective of the game master)
- Admin interface on the game master

# Setup:

The frontend is developed with react the backend uses socket io for client to client communication.
Run server.ts for nodejs server + express for frontend delivery
`node server`
The server will listen on port 3000
If frontend has been build `yarn build` the server uses the express framework to deliver the app build for production.

## React development

In the project directory, you can run:

### `yarn develop`

Runs the app in the development mode.<br />
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

The page will reload if you make edits.<br />
You will also see any lint errors in the console.

### `yarn start`

Runs the app as node server in production mode.<br />
Open [http://localhost:8080](http://localhost:8080) to view it in the browser.

### `yarn test`

Launches the test runner in the interactive watch mode.<br />
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `yarn build`

Builds the app for production to the `build` folder.<br />
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.<br />
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

### `yarn eject`

**Note: this is a one-way operation. Once you `eject`, you can’t go back!**

## Run using Docker

To build this project as docker container run the following command

```bash
docker build -t local/spotiguess:latest .
```

To run the image on port 80 use the following command:

```bash
docker run -d -p 80:8080 --name spotiguess local/spotiguess:latest
```


CHANGELOG.md
# Changelog

## Version 0.4.0

- Add local storage to remember current room and name for fast rejoin on connection lost
- Add local storage for last voting to keep the vote intact
- Improve deployment setup
- Fix open vulnerabilities

## Version 0.3.0

- Add about section
- Change configuration for development page
- Internal refactoring

## Version 0.2.0

- Switch to typescript to get more reliable source code
- Add separate query to fetch proper display name from api

## Version 0.1.0

- Release before migration to typescript


server.js
var express = require('express');
var fs = require("fs");
var app = express();
var http = require('http').createServer(app);
let config = {}

if (fs.existsSync('.env')) {
  config = {
    cors: {
      origin: "http://localhost:3000",
      methods: ["GET", "POST"]
    }
  }
  console.log("Development mode")
}

var io = require('socket.io')(http, config);
const port = process.env.PORT || 8080;

app.use('/',
        express.static(__dirname + '/build'));

app.get('/', function(req, res){
  res.sendFile(__dirname + '/build/' + 'index.html');
});

io.on('connection', function(socket){
  console.log('connected');

  socket.on('room-code', function(roomcode) {
    console.log('A new server has been created, code:' + roomcode);
    socket.join(roomcode + '_host');
  });

  socket.on('room-code-reconnect', function(roomcode) {
    console.log('A server has reconnected, code:' + roomcode);
    socket.join(roomcode + '_host');
  });

  // msg: {roomcode:str, name:str}
  socket.on('request-join', function(msg){
    socket.leaveAll();
    console.log('Join request: ' + JSON.stringify(msg));
    //Server side validation
    if (! /^[a-z][-a-z _0-9]{0,15}$/i.test(msg.name)) {
      console.log('rejected name');
      return;
    }
    if (! /^[A-Z]{5}$/.test(msg.roomcode)) {
      console.log('rejected code');
      return;
    }
    socket.join(msg.roomcode);
    socket.to(msg.roomcode + '_host').emit('request-join', msg);
  });

  // msg: {roomcode:str, name:str}
  socket.on('join-accepted', function(msg){
    socket.to(msg.roomcode).emit('join-accepted', msg.name);
  });

  // msg: {roomcode:str, options:Set}
  socket.on('options', function(msg){
    console.log('options for room: ' + JSON.stringify(msg));
    socket.to(msg.roomcode).emit('options', msg.options);
  });

  // msg: {name:str, roomcode:str, option:str}
  socket.on('vote', function(msg){
    console.log('Player voted: ' + JSON.stringify(msg));
    socket.to(msg.roomcode + '_host').emit('vote', msg);
  });

  // msg: {roomcode:str, artist:str, title:str, voteTime:num}
  socket.on('next-song', function(msg){
    console.log('next-song: ' + JSON.stringify(msg));
    socket.to(msg.roomcode).emit('next-song', msg);
  });

  socket.on('disconnect', function(){
    console.log('disconnected');
  });
});

http.listen(port, function(){
  console.log('listening on *:' + port);
});
